Documentation for managing a testing grid using Tripal Devtools (#137)
tripal/tripal_doc: docs/dev_guide/testing.rst

   testing/chadoTestTrait
   testing/fields

Automated testing across many Drupal/PHP Versions
------------------------------------------------------
After setting up automated testing with PHPUnit, it is considered best practice to run these tests across all supported Drupal and PHP versions for your module. This guide explains how to configure this using GitHub Workflows, the Tripal Test GitHub Action, and Tripal Devtools.

Creating a GitHub Workflow to test a number of Drupal - PHP - PostgreSQL combinations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
1. Copy `.github/workflows/ALL-phpunit.yml <https://github.com/tripal/tripal/blob/4.x/.github/workflows/ALL-phpunit.yml>`_ into your module.

> 2. Update the `directory-name <https://github.com/tripal/tripal/blob/4.x/.github/workflows/ALL-phpunit.yml#L84>`_ to match the name of the directory in ``/var/www/drupal/web/modules/contrib`` that your module code will be in.

Reviewer: Consider adding a phrase to say to open the file and modify the 2 yml properties directory-name and modules mentioned.

> 3. If your Dockerfile already enables the modules then `modules <https://github.com/tripal/tripal/blob/4.x/.github/workflows/ALL-phpunit.yml#L85>`_ can be left empty. If not, you can list the modules to be installed here.

Reviewer: Indicate what delimiter to use to separate each module if listing the modules.

And that's it! For more information about the parameters, see the `Tripal Test GitHub Action <https://github.com/tripal/test-tripal-action>`_ documentation.

> Generating a testing grid for your readme to show the current compatibility

Reviewer: Consider using all caps for readme - README to make it consistent to the reference made by command in the second screenshot (grid)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
1. Install tripal devtools in your container using the following commands (in the future this will be part of TripalDocker by default):

>   - Within the drupal root of the docker container (i.e. ``/var/www/drupal/web``), run the command ``composer require tripal/tripal_devtools``

Reviewer: Consider describing what the command does. For example, composer require will add the package to your project - run the command ..... to add the package to your project.

>   - Once it's installed correctly, run ``drush en tripal_devtools``

Reviewer: Similar to my previous comment - run the command .... to enable the module.

>   - Within the drupal root of the docker container (i.e. ``/var/www/drupal/web``), run the command ``composer require tripal/tripal_devtools`` to install tripal_devtools module in your docker container.

Reviewer 2: ```suggestion
  - Within the drupal root of the docker container (i.e. ``/var/www/drupal/web``), run the command ``composer require tripal/tripal_devtools`` to install the tripal_devtools module in your docker container.
```

2. Run the following command ``drush generate tripal-admin:readme-grid`` and answer the prompts. This will generate a number of Github workflows where each one runs a single combination from the matrix on the `4.x` branch when a PR is merged. This is used to generate the badges you will include in your readme.

.. image:: images/generate_grid_1.png

3. The command in step 2 will also print out the testing grid which you can copy and paste into your readme. See `the following for an example of this in Tripal <https://github.com/tripal/tripal/blob/4.x/README.md?plain=1#L11-L33>`_.

.. image:: images/generate_grid_2.png

When maintaining this grid, simply update the `.github/workflows/ALL-phpunit.yml <https://github.com/tripal/tripal/blob/4.x/.github/workflows/ALL-phpunit.yml>`_ with the new combinations you want to support, delete the workflow files and run the command again to re-create them and the associated grid.

Multiple versions of PHPUnit
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> When testing across multiple versions of PHP and Drupal you will likely find that different versions of PHPUnit are being used in some combinations. For example, Drupal 10.4 and 10.5 utilize PHPUnit 9, Drupal 11.1 uses PHPUnit 10 and Drupal 11.2 and 11.3 use PHPUnit 11. Unfortunatly PHPUnit configuration is usually not backwards compatible between major versions.

Reviewer: A small typo -
```suggestion
When testing across multiple versions of PHP and Drupal you will likely find that different versions of PHPUnit are being used in some combinations. For example, Drupal 10.4 and 10.5 utilize PHPUnit 9, Drupal 11.1 uses PHPUnit 10 and Drupal 11.2 and 11.3 use PHPUnit 11. Unfortunately, PHPUnit configuration is usually not backwards compatible between major versions.
```

> Thankfully TripalTest Github Action handles this natively as long as you use a specific naming structure for your phpunit.xml. The docs say:

Reviewer: ```suggestion
Thankfully, TripalTest Github Action handles this natively as long as you use a specific naming structure for your phpunit.xml. The docs say:
```

> Thankfully TripalTest Github Action handles this natively as long as you use a specific naming structure for your phpunit.xml. The docs say:

Reviewer: ```suggestion
Thankfully TripalTest Github Action handles this natively as long as you use a specific naming structure for your phpunit.xml. The official PHPUnit documentations indicate:
```

::

    The version of phpunit will be checked in the docker container and the action will 
    try to detect the correct PHPUnit XML configuration file to use. The following 
    naming formats are supported: phpunit.MAJOR.MINOR.xml, phpunit.MAJOR.xml, 
    phpunit.xml where MAJOR = 10 and MINOR = 5 for PHPUnit 10.5. You can supply 
    the file to use via the phpunit-config option to override this.

::

To set this up, you will want to 

1. Start a docker container on a particular combination of versions to target one of the PHPUnit versions. For example, Drupal 10.5 and PHP 8.2.
2. Within the container, you will want to copy the default phpunit.xml configuration for that version of Drupal into your module root and name it based on the version. For example, ``cp /var/www/drupal/web/core/phpunit.xml.dist /var/www/drupal/web/modules/contrib/MYMODULE/phpunit.9.xml`` where your module code is in a directory named `MYMODULE` and the major version of PHPUnit is 9.
3. You will want to repeat the above two steps for other combinations targetting the remaining PHPUnit versions.

> 4. For each PHPUnit.xml you will want to configure your test files and coverage on the section of the xml file shown below. You can see the phpunit.xml configuration files in Tripal Core for an example of how to do this.

Reviewer: ```suggestion
4. For each PHPUnit.xml, configure your test files and coverage on the section of the xml file shown below. You can see the phpunit.xml configuration files in Tripal Core for an example of how to do this.
```
